presto-main/src/main/java/io/prestosql/server/DynamicFilterService.java

```diff
@@ -0,0 +1,272 @@
+                            }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    public ListenableFuture<TupleDomain<String>> getSummary(QueryId queryId, String source)
+    {
+        return dynamicFilterSummaries.computeIfAbsent(SourceDescriptor.of(queryId, source), ignored -> SettableFuture.create());
+    }
+
+    public void removeQuery(QueryId queryId)
+    {
+        dynamicFilterSummaries.keySet().removeIf(sourceDescriptor -> sourceDescriptor.getQueryId().equals(queryId));
+        queries.remove(queryId);
+    }
+
+    private static List<StageInfo> getStagesWithDynamicFilters(List<StageInfo> stages)
+    {
+        return stages.stream()
+                .filter(stageInfo -> PlanNodeSearcher
+                        .searchFrom(stageInfo.getPlan().getRoot())
+                        .where(planNode -> {
+                            if (planNode instanceof JoinNode) {
+                                JoinNode joinNode = (JoinNode) planNode;
+                                return !joinNode.getDynamicFilters().isEmpty();
+                            }
+                            return false;
+                        })
+                        .matches())
+                .collect(toImmutableList());
+    }
+
+    @Immutable
+    private static class SourceDescriptor
+    {
+        private final QueryId queryId;
+        private final String source;
+
+        public static SourceDescriptor of(QueryId queryId, String source)
+        {
+            return new SourceDescriptor(queryId, source);
+        }
+
+        private SourceDescriptor(QueryId queryId, String source)
+        {
+            this.queryId = requireNonNull(queryId, "queryId is null");
+            this.source = requireNonNull(source, "source is null");
+        }
+
+        public QueryId getQueryId()
+        {
+            return queryId;
+        }
+
+        public String getSource()
+        {
+            return source;
+        }
+
+        @Override
+        public boolean equals(Object other)
+        {
+            if (other == this) {
+                return true;
+            }
+            if (other == null || getClass() != other.getClass()) {
+                return false;
+            }
+
+            SourceDescriptor sourceDescriptor = (SourceDescriptor) other;
+
+            return Objects.equals(queryId, sourceDescriptor.queryId) &&
+                    Objects.equals(source, sourceDescriptor.source);
+        }
+
+        @Override
+        public int hashCode()
+        {
+            return Objects.hash(queryId, source);
+        }
+
+        @Override
+        public String toString()
+        {
+            return toStringHelper(this)
+                    .addValue(queryId)
+                    .addValue(source)
+                    .toString();
+        }
+    }
+
+    public static class DynamicFilterSupplier
```

Comment:
make this method in dynamic filter service instead.

Reply:
added `createDynamicFilterSupplier` now